text/0000-cortex-m-targets.md

```diff
@@ -0,0 +1,608 @@
+
+# Tell Cargo about the linker and the linker script.
+$ mkdir .cargo && edit .cargo/config && cat .cargo/config
+```
+
+``` toml
+[build]
+rustflags = ["-C", "link-args=-Tlayout.ld"]
+
+[target.cortex-m3]
+linker = "arm-none-eabi-gcc"
+```
+
+```
+# The is the linker script
+$ edit layout.ld && cat layout.ld
+/* Device-specific memory regions */
+MEMORY
+{
+  FLASH : ORIGIN = 0x08000000, LENGTH = 128K
+  SRAM  : ORIGIN = 0x20000000, LENGTH = 8K
+}
+
+/* Omitted:
+   - Definition and placement of the vector table and other program sections
+   - Addresses of memory mapped peripherals */
+
+$ edit src/main.rs && cat src/main.rs
+```
+
+``` rust
+#![no_main]
+#![no_std]
+
+// Program entry point
+fn start() -> ! {
+    // ...
+}
+
+// Omitted: definition of lang items, placing `start` in the vector table, etc.
+```
+
+```
+$ cargo build --target cortex-m3
+
+$ file target/cortex-m3/debug/app
+target/cortex-m3/debug/app ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, not stripped
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None so far.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Don't do this
+
+Leave it up to the user figure out how to set up their development environment.
+
+Luckily, the author is currently writing [extensive documentation] on the area of Rust on
+microcontrollers, and also has developed tools that make cross compiling [freestanding crates] and
+[compiler-rt] relatively easy. However, adding these targets to the compiler will make the setup
+process simpler (less steps, requires less tooling, etc).
+
+[extensive documentation]: http://japaric.github.io/copper
+[freestanding crates]: https://github.com/japaric/xargo
+[compiler-rt]: https://github.com/japaric/compiler-rt.rs
+
+## Use the hard-float CC instead of the soft-float one
+
+The main proposal opted for the soft-float CC, favoring interoperability over performance. However,
+using the hard-float CC to favor performance is also a valid alternative.
+
+Yet another alternative is to provide both CC variants for the targets that support them. For
+example, instead of a single `cortex-m4f` target now there would be a `cortex-m4f` target for the
+soft-float CC and a `cortex-m4f-hf` target for the hard-float CC.
+
+## Less targets
+
+It's possible to reduce the number of targets that need to be added if the CPU specific
+optimizations (the `cpu` field in the target specifications) are dropped. This leads to the
+following smaller set of targets:
+
+- `thumbv6m-none-eabi`: Subsumes the `cortex-m0`, `cortex-m0plus` and `cortex-m1` targets.
+- `thumbv7m-none-eabi`: Equivalent to the `cortex-m3` target.
+- `thumbv7em-none-eabi`: Subsumes the `cortex-m4` and `cortex-m7` targets. No FPU instructions.
+  Soft-float calling convention.
+- `thumbv7em-none-eabihf`: Like `thumbv7em-none-eabi` but uses the hard-float calling convention.
+  With FPU instructions; the FPU will be set to the lowest common denominator between the cortex-m4
```

Comment:
I think this would actually be best simply because that is the current convention in the curent list of targets.
```
#rustc --print target-list
aarch64-apple-ios
aarch64-linux-android
aarch64-unknown-linux-gnu
arm-linux-androideabi
arm-unknown-linux-gnueabi
arm-unknown-linux-gnueabihf                                                    
armv7-apple-ios
armv7-linux-androideabi
armv7-unknown-linux-gnueabihf
armv7s-apple-ios  
```
I don't recommend deviating from this convention.